src/main.cpp

```diff
@@ -981,6 +981,36 @@ bool ContextualCheckTransaction(
             // https://zips.z.cash/zip-0213#specification
             uint256 ovk;
             for (const OutputDescription &output : tx.vShieldedOutput) {
+                bool zip_212_enabled;
+                libzcash::SaplingPaymentAddress zaddr;
+                CAmount value;
+
+            // EoS height for 5.3.3 and 5.4.2 is 2121024.
+            if (nHeight >= 2121200) {
+                try {
+                    auto decrypted = wallet::try_sapling_output_recovery(
+                        *chainparams.RustNetwork(),
+                        nHeight,
+                        ovk.GetRawBytes(),
+                        {
+                            output.cv.GetRawBytes(),
+                            output.cmu.GetRawBytes(),
+                            output.ephemeralKey.GetRawBytes(),
+                            output.encCiphertext,
+                            output.outCiphertext,
+                        });
+                    zip_212_enabled = decrypted->zip_212_enabled();
+
+                    libzcash::SaplingNotePlaintext notePt;
+                    std::tie(notePt, zaddr) = SaplingNotePlaintext::from_rust(std::move(decrypted));
+                    value = notePt.value();
```

Comment:
Should there be an `assert(notePt.has_value());` here? (If `notePt.value()` throws `std::bad_optional_access`, that won't be caught by the `catch` below.)

Reply:
This is the value of the note, not the value of a `std::optional`.

Reply:
Oh, right.